video/out/vo_sixel.c

```diff
@@ -0,0 +1,429 @@
+    priv->image_height = params->h;
+    priv->image_width  = params->w;
+    priv->image_format = params->imgfmt;
+    save_image_offset_location(vo);
+
+    priv->sws->src = *params;
+    priv->sws->dst = (struct mp_image_params) {
+        .imgfmt = IMGFMT,
+        .w = priv->opts->width,
+        .h = priv->opts->height,
+        .p_w = 1,
+        .p_h = 1,
+    };
+
+    priv->frame = mp_image_alloc(IMGFMT, priv->opts->width, priv->opts->height);
+    if (!priv->frame)
+        return -1;
+
+    if (mp_sws_reinit(priv->sws) < 0)
+        return -1;
+
+    printf(ESC_HIDE_CURSOR);
+    printf(ESC_CLEAR_SCREEN);
+    vo->want_redraw = true;
+
+    return resize(vo);
+}
+
+static void draw_image(struct vo *vo, mp_image_t *mpi)
+{
+    struct priv *priv = vo->priv;
+    struct mp_image src = *mpi;
+
+    // Downscale the image
+    mp_sws_scale(priv->sws, priv->frame, &src);
+
+    // Copy from mpv to RGB format as required by libsixel
+    memcpy_pic(priv->buffer, priv->frame->planes[0], priv->opts->width * depth, priv->opts->height,
+               priv->opts->width * depth, priv->frame->stride[0]);
+
+    if (priv->opts->fixedpal)
+        prepare_static_palette(vo);
+    else
+        prepare_dynamic_palette(vo);
+
+    talloc_free(mpi);
+}
+
+static int sixel_write(char *data, int size, void *priv)
+{
+    return fwrite(data, 1, size, (FILE *)priv);
+}
+
+static void flip_page(struct vo *vo)
+{
+    struct priv* priv = vo->priv;
+
+    // Go to the offset row and column, then display the image
+    printf(ESC_RESTORE_SAVED_CURSOR);
+    sixel_encode(priv->buffer, priv->opts->width, priv->opts->height,
+                 PIXELFORMAT_RGB888,
+                 priv->dither, priv->output);
+    fflush(stdout);
+}
+
+static int preinit(struct vo *vo)
+{
+    struct priv *priv = vo->priv;
+    SIXELSTATUS status = SIXEL_FALSE;
+    FILE* sixel_output_file = stdout;
+
+    // Parse opts set by CLI or conf
+    priv->opts = mp_get_config_group(vo, vo->global, &vo_sixel_conf);
+    priv->sws = mp_sws_alloc(vo);
+    priv->sws->log = vo->log;
+    mp_sws_enable_cmdline_opts(priv->sws, vo->global);
+
+    status = sixel_output_new(&priv->output, sixel_write, sixel_output_file, NULL);
+    if (SIXEL_FAILED(status))
+        return status;
+
+    printf(ESC_HIDE_CURSOR);
+
+    /* don't use private color registers for each frame. */
+    printf(ESC_USE_GLOBAL_COLOR_REG);
+
+    priv->dither = NULL;
+    status = sixel_dither_new(&priv->testdither, priv->opts->reqcolors, NULL);
+
+    if (SIXEL_FAILED(status))
+        return status;
+
+    priv->buffer =
+        talloc_array(NULL, uint8_t, depth * priv->opts->width * priv->opts->height);
+
+    return 0;
+}
+
+static int query_format(struct vo *vo, int format)
+{
+    return format == IMGFMT;
+}
+
+static int control(struct vo *vo, uint32_t request, void *data)
+{
+    return VO_NOTIMPL;
+}
+
+
+static void uninit(struct vo *vo)
+{
+    struct priv *priv = vo->priv;
+
+    printf(ESC_TERMINATE_DCS);
```

Comment:
Where does the DCS sequence start? If that's the entire image in a DCS sequence, shouldn't it start and end on each frame?

Reply:
DCS I think is set by the sixel encode function. My guess is that if the player exits and sixel output is interrupted and then this function for cleanup gets called, it exits gracefully (could be wrong though)

Reply:
What do you mean "my guess"? did you try without it? Do you have a use case where using it makes a difference?

Reply:
It doesn't affect playback, only the cleanup part on exit. So playback will not be affected regardless of whether this line is removed or it exists.

Reply:
I wasn't asking about playback. I see it's called during uninit. I'm asking if you know of a use case where it makes a difference.

Reply:
I tested without it, and didn't cause any noticeable issue for me on xterm and mlterm, so I have removed it as per your suggestion.

Reply:
I didn't suggest to remove it. I asked you to understand why it's there, and in what use cases this termination helps with some things.

My guess is that at most it's useful in some abort conditions (crash or killed externally), but at such case I don't even know if uninit would be called.

Anyway, it's OK that you removed it - let's keep it that way, but if you can still figure out why it was there in the first place, it would be best.